liquid-staking/src/delegation.rs

```diff
@@ -0,0 +1,90 @@
+use multiversx_sc::imports::*;
+pub const UNBOND_PERIOD: u64 = 10;
+
+use crate::common::{self, storage::Epoch};
+
+#[multiversx_sc::module]
+pub trait DelegationModule: common::storage::CommonStorageModule {
+    #[payable("ELGD")]
+    #[endpoint(stake)]
+    fn stake(&self) {
+        let caller = self.blockchain().get_caller();
+        let egld_amount = self.call_value().egld_value().clone_value();
+        let delegation_contract_address = self.delegation_address().get();
+        let delegation_endpoint = ManagedBuffer::from("delegate");
+
+        self.tx()
+            .to(delegation_contract_address)
+            .raw_call(delegation_endpoint)
+            .egld(&egld_amount)
+            .with_callback(DelegationModule::callbacks(self).stake_callback(&caller, &egld_amount))
+            .call_and_exit();
+    }
+
+    #[endpoint(unStake)]
+    fn unstake(&self, egld_amount_to_unstake: BigUint) {
+        let caller = self.blockchain().get_caller();
+        let current_epoch = self.blockchain().get_block_epoch();
+        let total_egld_deposit = self.delegated_value(caller.clone()).get();
+        let delegation_contract_address = self.delegation_address().get();
+        let undelegate_endpoint = ManagedBuffer::from("unDelegate");
+
+        let mut args: ManagedArgBuffer<Self::Api> = ManagedArgBuffer::new();
+        args.push_arg(&egld_amount_to_unstake);
+
```

Comment:
I feel like it is way better to make a  check here if caller has something to unstake, by checking that the amount to unstake is available in `delegated_value`

in this way you will not call undelegate if you already know that the caller hasn't enough to unstake